File: euler-scala/src/net/projecteuler/level001/Problem0003.scala
package net.projecteuler.level001

import scala.math.BigInt
import java.math.BigInteger

/*
 * The prime factors of 13195 are 5, 7, 13 and 29.
 *
 * What is the largest prime factor of the number 600851475143 ?
 */
object Problem0003 {
  def main(args: Array[String]): Unit = {
    
    val n = 600851475143L
    
    def rec(n: Long, d: Long):Long = 
      if(n == d) d
      else {
        if(n % d == 0) rec(n/d, d)
        else rec(n,d + 1)
      }
      
    println(rec(n,2))
    
  }
}

File: euler-java/src/net/projecteuler/level001/Problem0003.java
package net.projecteuler.level001;

/*
 * The prime factors of 13195 are 5, 7, 13 and 29.
 *
 * What is the largest prime factor of the number 600851475143 ?
 */
public class Problem0003 {

  public static void main(String[] args) {

    long a = 600851475143L, b = 5;

    for (; b < a;) {

      if (a % b == 0) {

        a = a / b;

      } else {
        b++;
      }

    }

    System.out.println(b);

  }

}
